### PR TITLE
Avoid duplicates with metadata caching client on iterateStatus

### PR DESCRIPTION
**Pull request by apc999**
Fix https://github.com/Alluxio/alluxio/issues/12373

### CODE REVIEW

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 1]
```diff
@@ -96,17 +97,20 @@ public void iterateStatus(AlluxioURI path, ListStatusPOptions options,
       // Otherwise, needs more complicated logic inside the cache,
       // that might not worth the effort of caching.
       super.iterateStatus(path, options, action);
+      return;
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:100]: this was a bug which cause two listStatus RPCs on one iterateStatus API call

**Comment by gpang** [comment]
```
2020-10-29T22:54:30.5991025Z 22:54:30.594 [ERROR] /usr/src/alluxio/core/client/fs/src/test/java/alluxio/client/file/MockFileSystemMasterClient.java:[109,3] method does not override or implement a method from a supertype
```

**Comment by apc999** [comment]
@gpang  can you take another look?

File: core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java [hunk 2]
```diff
@@ -96,17 +97,20 @@ public void iterateStatus(AlluxioURI path, ListStatusPOptions options,
...
     }
 
-    List<URIStatus> statuses = mMetadataCache.listStatus(path);
-    if (statuses == null) {
+    List<URIStatus> cachedStatuses = mMetadataCache.listStatus(path);
+    if (cachedStatuses == null) {
+      List<URIStatus> statuses = new ArrayList<>();
       super.iterateStatus(path, options, status -> {
-        mMetadataCache.put(path, status);
```
Comment (apc999) [inline core/client/fs/src/main/java/alluxio/client/file/MetadataCachingBaseFileSystem.java:104]: `mMetadataCache.put(path, status);` is not caching `ls` result of a dir and it should be `status.getPath`, not `path`

**Review by gpang — APPROVED** [review]
LGTM

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by LuQQiu** [comment]
alluxio-bot, cherry-pick this to branch-2.4 please

**Pull request merged**